File: day17/puzzle.py
import re
from dataclasses import dataclass
from aoc import AocPuzzle


TEST_ANSWERS = {
    "part1": 45,
    "part2": 112
}


@dataclass
class TargetArea:
    min_x: int
    max_x: int
    min_y: int
    max_y: int


class Puzzle(AocPuzzle[TargetArea, int]):
    def __init__(self):
        super().__init__(TEST_ANSWERS)


    def parse_input(self, lines: list[str]):
        matches = re.search(r"x=(.+)\.\.(.+), y=(.+)\.\.(.+)", lines[0])
        return TargetArea(*[int(match) for match in matches.groups()])


    def get_dxs(self):
        dxs: list[int] = []
        dx = 1
        while dx < self._input.min_x:
            x = dx
            cur_dx = dx
            while x <= self._input.max_x:
                if x >= self._input.min_x:
                    dxs.append(dx)
                    break
                elif cur_dx == 1:
                    break

                cur_dx -= 1
                x += cur_dx

            dx += 1

        dxs.extend(range(self._input.min_x, self._input.max_x + 1))
        return dxs


    def get_dys(self):
        dys: list[int] = []
        dy = -1
        while dy > self._input.max_y:
            y = dy
            cur_dy = dy
            while y >= self._input.min_y:
                if y <= self._input.max_y:
                    dys.append(dy)
                    break

                cur_dy -= 1
                y += cur_dy

            dy -= 1

        dys.extend(range(self._input.max_y, self._input.min_y - 1, -1))
        dys.extend([-dy - 1 for dy in dys])
        dys.sort()
        return dys


    def run_part1(self):
        dys = self.get_dys()
        return (dys[-1] * (dys[-1] + 1)) // 2


    def run_part2(self):
        dxs = self.get_dxs()
        dys = self.get_dys()
        velocity_count = 0
        for dx in dxs:
            tmp = dx * (dx + 1)
            for dy in dys:
                if dy > 0:
                    cur_dx = max(dx - 2 * dy - 1, 0)
                    cur_dy = -dy - 1
                    x = (tmp - cur_dx * (cur_dx + 1)) // 2
                    y = 0
                else:
                    cur_dx = max(dx - 1, 0)
                    cur_dy = dy - 1
                    x = dx
                    y = dy

                while x <= self._input.max_x and y >= self._input.min_y:
                    if x >= self._input.min_x and y <= self._input.max_y:
                        velocity_count += 1
                        break

                    x += cur_dx
                    y += cur_dy
                    cur_dy -= 1
                    if cur_dx > 0:
                        cur_dx -= 1

        return velocity_count

File: day22/puzzle.js
import { createAocPuzzle } from "./aoc.js";

const testAnswers = {
  part1: 590784,
  part2: 2758514936282235
};

function parseInput(lines) {
  const list = [];
  let rebootSteps = [];
  for (const line of lines) {
    if (!line.trim()) {
      list.push(rebootSteps);
      rebootSteps = [];
    } else {
      const match = /^(.+?) x=(.+?)\.\.(.+?),y=(.+?)\.\.(.+?),z=(.+?)\.\.(.+?)$/.exec(line);
      rebootSteps.push({
        on: (match[1] == "on"),
        minX: +match[2],
        maxX: +match[3],
        minY: +match[4],
        maxY: +match[5],
        minZ: +match[6],
        maxZ: +match[7]
      });
    }
  }
  list.push(rebootSteps);
  return list;
}

function getCutOutPlanes(basePlane, otherPlane) {
  if (
    basePlane.minX > otherPlane.maxX ||
    basePlane.maxX < otherPlane.minX ||
    basePlane.minY > otherPlane.maxY ||
    basePlane.maxY < otherPlane.minY
  ) {
    return null;
  }

  const planes = [];
  const minY = Math.max(basePlane.minY, otherPlane.minY);
  const maxY = Math.min(basePlane.maxY, otherPlane.maxY);
  if (otherPlane.minY > basePlane.minY) {
    planes.push({ minX: basePlane.minX, maxX: basePlane.maxX, minY: basePlane.minY, maxY: minY - 1 });
  }
  if (otherPlane.minX > basePlane.minX) {
    planes.push({ minX: basePlane.minX, maxX: otherPlane.minX - 1, minY, maxY });
  }
  if (otherPlane.maxX < basePlane.maxX) {
    planes.push({ minX: otherPlane.maxX + 1, maxX: basePlane.maxX, minY, maxY });
  }
  if (otherPlane.maxY < basePlane.maxY) {
    planes.push({ minX: basePlane.minX, maxX: basePlane.maxX, minY: maxY + 1, maxY: basePlane.maxY });
  }

  return planes;
}

function getOnCount(rebootSteps) {
  let onCount = 0;
  const minZ = Math.min(...rebootSteps.map(step => step.minZ));
  const maxZ = Math.max(...rebootSteps.map(step => step.maxZ));

  for (let z = minZ; z <= maxZ; ++z) {
    const onPlanes = [];
    steploop: for (const step of rebootSteps) {
      if (step.minZ > z || step.maxZ < z) {
        continue;
      }

      if (step.on) {
        const curPlanes = [step];
        for (const onPlane of onPlanes) {
          for (let i = 0; i < curPlanes.length; ++i) {
            const newPlanes = getCutOutPlanes(curPlanes[i], onPlane);
            if (newPlanes) {
              curPlanes.splice(i, 1, ...newPlanes);
              if (!curPlanes.length) {
                continue steploop;
              }
              i += newPlanes.length - 1;
            }
          }
        }
        onPlanes.push(...curPlanes);
      } else {
        for (let i = 0; i < onPlanes.length; ++i) {
          const newPlanes = getCutOutPlanes(onPlanes[i], step);
          if (newPlanes) {
            onPlanes.splice(i, 1, ...newPlanes);
            i += newPlanes.length - 1;
          }
        }
      }
    }

    for (const plane of onPlanes) {
      onCount += (plane.maxX - plane.minX + 1) * (plane.maxY - plane.minY + 1);
    }
  }

  return onCount;
}

function runPart1(input) {
  const steps = input[0].filter(step => (
    step.minX >= -50 && step.maxX <= 50 &&
    step.minY >= -50 && step.maxY <= 50 &&
    step.minZ >= -50 && step.maxZ <= 50
  ));
  return getOnCount(steps);
}

function runPart2(input, isTest = false) {
  return getOnCount(input[isTest ? 1 : 0]);
}

export function createPuzzle() {
  return createAocPuzzle({ parseInput, runPart1, runPart2, testAnswers });
}
